### DISCUSSION TITLE
Meaningful names for Action/Func event parameters

### DISCUSSION THREAD

**Discussion by IS4Code** [dotnet/csharplang#5706] [Language Ideas]
### The issue

Oftentimes I see a library that uses "simple" events, i.e. in this form:
```cs
public event Action<int, string, bool> SomethingDone;
```
This is very convenient from the implementor's perspective, when compared to the standard *EventHandler*/*EventArgs* approach, but from the user's perspective, it doesn't convey any information about what these parameters actually mean. This becomes irritating when an auto-generated method for the event gets parameters like `arg0`, `arg1` etc., and there is no part of the structured documentation that would contain this information.

It's especially inconvenient in a case like this:
```cs
public event Action<int, int> ValueChanged;
```
Is it `oldValue`, `newValue`, or `newValue`, `oldValue`? Can't tell.

I am aware that the "proper" solution would be to create a meaningful delegate for the event, i.e.
```cs
public delegate void ValueChangedDelegate(int oldValue, int newValue);
public event ValueChangedDelegate ValueChanged;
```
But this is too long to write and thus not so attractive when compared to the simple but confusing solution.

### The solutions

I have two solutions to this issue. First, there could be a syntax that auto-generates the delegate at the event definition site, since most of the times, a delegate is used only once for a specific event anyway:
```cs
public event void ValueChanged(int oldValue, int newValue);
```
This would be a shorthand for the snippet above. It has the advantage of having a meaningful delegate type for the handler, so you can carry it around without it getting mixed with other delegate types.

A second solution is inspired by tuples. What I can already do is to treat all the parameters as a tuple and give them names this way:
```cs
public event Action<(int oldValue, int newValue)> ValueChanged;
```
If custom attributes make it possible to preserve the names, why not get rid of the tuple altogether?
```cs
public event Action<int oldValue, int newValue> ValueChanged;
```

**Comment by PathogenDavid** [comment]
See https://github.com/dotnet/csharplang/discussions/269 and https://github.com/dotnet/csharplang/discussions/1883

Also related: https://github.com/dotnet/csharplang/discussions/470

**Comment by TahirAhmadov** [comment]
I like the first option; it should create a strongly-typed anonymous delegate type.

**Comment by jez9999** [comment]
The first option doesn't have the convenience of allowing the callback to be declared inline in a method signature though, does it?  Being able to say `int foo(Func<T, T, T> fnCallback) {}` is very nice IMO.

I'd support the second option, *or* allowing shorthand for single-value tuples which would just allow use to use the tuple approach all the time.  That might at least be a "quick fix" until Func type arguments allow the specification of names, which seems like a more significant change.
